lib/puppet/http/client.rb

```diff
@@ -0,0 +1,210 @@
+      end
+    end
+  rescue Puppet::HTTP::HTTPError
+    raise
+  rescue => e
+    raise Puppet::HTTP::ConnectionError.new(_("Failed to connect to %{uri}: %{message}") % {uri: uri, message: e.message}, e)
+  end
+
+  def get(url, headers: {}, params: {}, ssl_context: nil, user: nil, password: nil, &block)
+    query = encode_params(params)
+    unless query.empty?
+      url = url.dup
+      url.query = query
+    end
+
+    request = Net::HTTP::Get.new(url, @default_headers.merge(headers))
+
+    apply_auth(request, user, password)
+
+    if block_given?
+      execute_streaming(request, ssl_context: ssl_context) do |response|
+        yield response
+      end
+    else
+      execute(request, ssl_context: ssl_context)
+    end
+  end
+
+  def put(url, headers: {}, params: {}, content_type:, body:, ssl_context: nil, user: nil, password: nil)
+    query = encode_params(params)
+    unless query.empty?
+      url = url.dup
+      url.query = query
+    end
+
+    request = Net::HTTP::Put.new(url, @default_headers.merge(headers))
+    request.body = body
+    request['Content-Length'] = body.bytesize
+    request['Content-Type'] = content_type
+
+    apply_auth(request, user, password)
+
+    execute(request, ssl_context: ssl_context)
+  end
+
+  def close
+    @pool.close
+  end
+
+  def default_ssl_context
+    unless @default_ssl_context
+      begin
+        cert = Puppet::X509::CertProvider.new
+        password = cert.load_private_key_password
+        ssl = Puppet::SSL::SSLProvider.new
+        @default_ssl_context = ssl.load_context(certname: Puppet[:certname], password: password)
+      rescue => e
+        # TRANSLATORS: `message` is an already translated string of why SSL failed to initialize
+        Puppet.log_exception(e, _("Failed to initialize SSL: %{message}") % { message: e.message })
+        # TRANSLATORS: `puppet agent -t` is a command and should not be translated
+        Puppet.err(_("Run `puppet agent -t`"))
+        raise e
+      end
+    end
+    @default_ssl_context
+  end
+
+  private
+
+  def execute(request, ssl_context:)
+    redirects = 0
+    retries = 0
+
+    loop do
+      connect(request.uri, ssl_context: ssl_context) do |http|
+        response = Puppet::HTTP::Response.new(http.request(request))
+        begin
+          request_callback(request, response)
+
+          if @redirector.redirect?(request, response)
+            request = @redirector.redirect_to(request, response, redirects)
+            redirects += 1
+            next
+          elsif @retry_after_handler.retry_after?(request, response)
+            interval = @retry_after_handler.retry_after_interval(request, response, retries)
+            retries += 1
+            if interval
+              Puppet.warning(_("Sleeping for %{interval} seconds before retrying the request") % { interval: interval })
+              ::Kernel.sleep(interval)
+              next
+            end
+          end
+        ensure
+          response.drain
+        end
```

Comment:
The streaming and non-streaming execute are very similar. Look at refactoring that.

Reply:
Done